src/mcts/search.cc

```diff
@@ -135,56 +144,72 @@ void ApplyDirichletNoise(Node* node, float eps, double alpha) {
-  std::tuple<int, float, float> best(-1, 0.0, 0.0);
-  for (Node* node : parent->Children()) {
-    if (parent == root_node_ && !limits_.searchmoves.empty() &&
-        std::find(limits_.searchmoves.begin(), limits_.searchmoves.end(),
-                  node->GetMove()) == limits_.searchmoves.end()) {
-      continue;
-    }
-    std::tuple<int, float, float> val(node->GetN(), node->GetQ(-10.0, 0.0),
-                                      node->GetP());
-    if (val > best) {
-      best = val;
-      best_node = node;
+float Search::CalculateTemperature() const {
+  // Put a floor on temperature to limit powers computed. Temps below the floor
+  // are rounded to 0. We round each component separately for better control.
+  float floor = 0.10f; // semi-arbitrary
+
+  // movetemp ~ 1/log(move*decay)
+  int move = played_history_.Last().GetGamePly() / 2;
+  float movetemp = kMoveTemp / (1 + std::log(1 + kMoveTempDecay * move));
+  if (movetemp < floor) movetemp = 0.0f;
+
+  // evaltemp ~ 1/log(eval^4*decay)
+  float eval = -root_node_->GetQ(0, 0);
+  float evaltemp = kEvalTemp /
+                   (1 + std::log(1 + kEvalTempDecay * std::pow(eval, 4)));
+  if (evaltemp < floor) evaltemp = 0.0f;
+
+  float temperature = movetemp + evaltemp;
+  //std::cout << "move: " << move << " movetemp: " << movetemp << " eval: " << eval << " evaltemp: " << evaltemp << " temp: " << temperature << "\n";
+
+  return temperature;
+}
+
+// Returns a child with most visits, if temp==0, or child chosen randomly
+// according to (visits_fraction)^(1/temp)
+Node* Search::GetBestChild(Node* parent,
+                           float temperature) const {
+
+  if (temperature <= 0.0f || parent->GetN() <= 1) {
+    // temp shouldn't be < 0 ever, but whatever
+    Node* best_node = nullptr;
+    // break visit count ties with search-eval, and secondarily with policy
+    std::tuple<int, float, float> best(-1, 0.0f, 0.0f);
+    for (Node* node : parent->Children()) {
+      if (parent == root_node_ && NodeInSearchmoves(node)) {
+        continue;
+      }
+      std::tuple<int, float, float> val(node->GetN(), node->GetQ(-10.0f, 0.0f),
+                                        node->GetP());
+      if (val > best) {
+        best = val;
+        best_node = node;
+      }
     }
+    return best_node;
   }
-  return best_node;
-}
 
-Node* Search::GetBestChildWithTemperature(Node* parent,
-                                          float temperature) const {
-  std::vector<float> cumulative_sums;
-  float sum = 0.0;
+  float coldness = 1 / temperature;
+  std::vector<float> partial_sums;
+  float sum = 0.0f;
   const float n_parent = parent->GetN();
 
   for (Node* node : parent->Children()) {
-    if (parent == root_node_ && !limits_.searchmoves.empty() &&
-        std::find(limits_.searchmoves.begin(), limits_.searchmoves.end(),
-                  node->GetMove()) == limits_.searchmoves.end()) {
+    if (parent == root_node_ && NodeInSearchmoves(node)) {
```

Comment:
I would add comment explaining what's going on here. Something like "If we are in UCI searchmoves mode, don't look into moves not listed there".

Reply:
This should have been addressed in the PR that created that code then :P